scene/resources/style_box.cpp

```diff
@@ -316,67 +333,239 @@ bool StyleBoxFlat::get_border_blend() const {
-	vs->canvas_item_add_rect(p_canvas_item, Rect2(Point2i(r_add.pos.x - additional_border_size[MARGIN_LEFT], r_add.pos.y), Size2(additional_border_size[MARGIN_LEFT], r_add.size.height)), light_color);
-	vs->canvas_item_add_rect(p_canvas_item, Rect2(Point2i(r_add.pos.x + r_add.size.width, r_add.pos.y), Size2(additional_border_size[MARGIN_RIGHT], r_add.size.height)), dark_color);
-	vs->canvas_item_add_rect(p_canvas_item, Rect2(Point2i(r_add.pos.x - additional_border_size[MARGIN_LEFT], r_add.pos.y + r_add.size.height), Size2(r_add.size.width + additional_border_size[MARGIN_LEFT] + additional_border_size[MARGIN_RIGHT], additional_border_size[MARGIN_BOTTOM])), dark_color);
+	PoolVector2Array corners = PoolVector2Array();
+	corners.append(tl);
+	corners.append(tr);
+	corners.append(br);
+	corners.append(bl);
+
+	//draw corners
+	for (int i = 0; i < 4; i++) {
+
+		Color col = col_top;
+		if (i > 1)
+			col = col_bottom;
+
+		if (filled) {
+			//the circle has to be drawn twice, otherwise the circles for the corners are not visible
+			//super buggy ;)
+			vs->canvas_item_add_circle(p_canvas_item, corners[i], (float)corner_radius[i], col);
+			vs->canvas_item_add_circle(p_canvas_item, corners[i], (float)corner_radius[i], col);
+		} else {
+			for (float j = 0; j < border; j += 1) {
+				draw_arc(vs, p_canvas_item, corners[i], i, (float)corner_radius[i] - j, col);
+			}
+		}
+	}
+
+	//setup infill values
+	int border_left_right = border;
+	int border_top_bottom = border;
+	if (filled) {
+		border_top_bottom = rect.size.height / 2;
+		border_left_right = rect.size.width / 2;
+	}
+	Rect2 rect_top = Rect2(rect.pos.x + corner_radius[0], rect.pos.y, rect.size.width - corner_radius[0] - corner_radius[1], border_top_bottom);
+	Rect2 rect_bottom = Rect2(rect.pos.x + corner_radius[3], rect.pos.y + rect.size.height - border_top_bottom, rect.size.width - corner_radius[3] - corner_radius[2], border_top_bottom);
+	Rect2 rect_left = Rect2(rect.pos.x, rect.pos.y + corner_radius[0], border_left_right, rect.size.height - corner_radius[0] - corner_radius[3]);
+	Rect2 rect_right = Rect2(rect.pos.x + rect.size.width - border_left_right, rect.pos.y + corner_radius[1], border_left_right, rect.size.height - corner_radius[1] - corner_radius[2]);
+
+	vs->canvas_item_add_rect(p_canvas_item, rect_top, col_top);
+	vs->canvas_item_add_rect(p_canvas_item, rect_left, col_left);
+	vs->canvas_item_add_rect(p_canvas_item, rect_right, col_right);
+	//has to be drawn last
+	vs->canvas_item_add_rect(p_canvas_item, rect_bottom, col_bottom);
+}
+inline int *get_offset_corner_radius(int offset, const int *corner_radius) {
+	int inner_corner_radius[4];
+	for (int i = 0; i < 4; i++) {
+		int rad = corner_radius[i] - offset;
+		if (rad < 0)
+			rad = 0;
+		inner_corner_radius[i] = rad;
+	}
+	int *p = &inner_corner_radius[0];
```

Comment:
well it's not that good either, you are allocating an array everytime. I don't understand why you created this function in the first place tbh, I feel like you don't really need it